src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+-- By choice of `n'` this coloring has a monochromatic line `l'` with color class `C'`, where
+-- `C'` is a `k`-coloring of `n → m`.
+  obtain ⟨l', C', hl'⟩ := hn',
+-- If `C'` has a monochromatic line, then so does `C`. We will use this twice.
+  have mono_of_mono : (∃ l, is_mono C' l) → (∃ l, is_mono C l),
+  { rintro ⟨l, c, hl⟩,
+    refine ⟨l.horizontal (some ∘ l' (classical.arbitrary m)), c, λ x, _⟩,
+    rw [line.horizontal_apply, ←hl, ←hl'], },
+-- By choice of `n`, `C'` either has `r` color focused lines or a monochromatic line.
+  specialize hn C',
+  rcases hn with ⟨s, sr⟩ | _,
+-- By above, we are done if `C'` has a monochromatic line.
+  work_on_goal 1 { exact or.inr (mono_of_mono hn) },
+-- Here we assume `C'` has `r` color focused lines. We split into cases depending on whether one of
+-- these `r` lines has the same color as the focus point. -/
+  by_cases h : ∃ p ∈ s.lines, (p : almost_mono _).color = C' s.focus,
+-- If so then this is a `C'`-monochromatic line and we are done.
+  { obtain ⟨p, p_mem, hp⟩ := h,
+    refine or.inr (mono_of_mono ⟨p.line, p.color, _⟩),
+    rintro (_ | _), rw [hp, s.is_focused p p_mem], apply p.has_color, },
+-- If not, we get `r+1` color focused lines by taking the product of the `r` lines with `l'` and
+-- adding to this the vertical line obtained by the focus point and `l`.
+  refine or.inl ⟨⟨(s.lines.map _).cons ⟨(l'.map some).vertical s.focus, C' s.focus, λ x, _⟩,
+    sum.elim s.focus (l'.map some none), _, _⟩, _⟩,
+-- The vertical line is almost monochromatic.
+  { rw [vertical_apply, ←congr_fun (hl' x), line.map_apply], },
+  { refine λ p, ⟨p.line.prod (l'.map some), p.color, λ x, _⟩,
+-- The product lines are almost monochromatic.
+    rw [line.prod_apply, line.map_apply, ←p.has_color, ←congr_fun (hl' x)], },
+-- Our `r+1` lines have the same endpoint.
+  { simp_rw [multiset.mem_cons, multiset.mem_map],
+    rintros _ (rfl | ⟨q, hq, rfl⟩),
+    { rw line.vertical_apply, },
+    { rw [line.prod_apply, s.is_focused q hq], }, },
+-- Our `r+1` lines have distinct colors (this is why we needed to split into cases above).
+  { rw [multiset.map_cons, multiset.map_map, multiset.nodup_cons, multiset.mem_map],
+    exact ⟨λ ⟨q, hq, he⟩, h ⟨q, hq, he⟩, s.distinct_colors⟩, },
+-- Finally, we really do have `r+1` lines!
+  { rw [multiset.card_cons, multiset.card_map, sr], },
+end
+
+end line
+
+/-- A generalization of Van der Waerden's theorem: if `M` is a finitely colored commutative
+monoid, and `S` is a finite subset, then there exists a monochromatic homothetic copy of `S`.
+Formally, instead of taking `S` to be a subset we take it to be a finite type with a function
+`S → M`.
+
+This theorem is equivalent to Gallai's higher-dimensional version of Van der Waerden for the
+monoids `ℕ^d`.
+-/
+theorem exists_mono_homothetic_copy (M) (S : Type) (f : S → M) [fintype S] [add_comm_monoid M]
+  (k : Type) [fintype k] (C : M → k) :
+  ∃ (a : pnat) (b : M) (c : k), ∀ s, C ((a : ℕ) • (f s) + b) = c :=
```

Comment:
`pnat`s have much smaller library support than naturals with the positivity assumption, could you add a variant of this lemma for that as well (almost certainly it should have a very short proof using this version)?

Reply:
I guess I can just use `a > 0` instead of `a : pnat` if the former is preferred?